File: flutter/craftd_widget/lib/presentation/builder/craftd_builder_manager.dart
import '../ui/button/craftd_button_builder.dart';
import '../ui/empty/craftd_empty_builder.dart';
import '../ui/text/craftd_text_builder.dart';
import 'craftd_builder.dart';

class CraftDBuilderManager {
  final Map<String, CraftDBuilder> _builders = {
    CraftDButtonBuilder.keyBuilder: CraftDButtonBuilder(),
    CraftDTextBuilder.keyBuilder: CraftDTextBuilder(),
  };

  final Map<String, CraftDBuilder> customBuilders = {};

  CraftDBuilderManager add(List<CraftDBuilder> arrayCraftDBuilder) {
    for (var it in arrayCraftDBuilder) {
      _builders[it.key] = it;
  }
    return this;
  }

  CraftDBuilder getBuilder(String key) {
    if (customBuilders.isNotEmpty) {
      _builders.addAll(customBuilders);
    }
    return _builders[key] != null ? _builders[key]! : CraftDEmptyBuilder();
  }
}

File: ios/craftd-swiftui/Sources/craftd-swiftui/presentation/builder/CraftDBuilderManager.swift
import Foundation

public class CraftDBuilderManager {
    public init() {
        add(builders: [
            CraftDTextBuilder(),
            CraftDButtonBuilder(),
            CraftDCheckBoxBuilder(),
            CraftDProgressBarBuilder(),
        ])
    }
    
    private var listCraftDBuilder: [String: CraftDBuilder] = [:]
    
    public func add(builders: [CraftDBuilder]) -> CraftDBuilderManager {
        builders.forEach { builder in add(builder: builder) }
        return self
    }
    
    public func add(builder: CraftDBuilder) -> CraftDBuilderManager {
        listCraftDBuilder[builder.key] = builder
        return self
    }
    
    public func getBuilder(key: String) -> CraftDBuilder {
        guard let builder = listCraftDBuilder[key] else {
            return CraftDEmptyBuilder()
        }
        return builder
    }
}
